### PR TITLE
M2: LoRA training on MPS (Apple Silicon)

### PR DESCRIPTION
**Pull request by kwiksher**
Adds Apple Silicon (MPS) support to the LoRA trainer, so character / environment
 / prop LoRAs can be trained on Macs without CUDA.

   ## What changed

   - **`real_trainer.py`** — `RealLoraTrainer.is_available()` now accepts MPS as a
 valid accelerator (CUDA **or** Apple MPS), and the trainer driver sets
 `PYTORCH_ENABLE_MPS_FALLBACK=1` so unsupported MPS ops degrade to CPU instead of
 hard-failing.
   - **`run_zimage_trainer.py`** (Z-Image backend) — accelerator resolved as CUDA >
 MPS > CPU; bf16 detection for MPS; unified-memory budget via
 `torch.mps.recommended_max_memory()`; MPS OOM surfaced as an OOM-style failure.
   - **`run_trainer.py`** (SDXL-legacy backend) — same CUDA > MPS > CPU resolution;
 bf16 mixed precision is CUDA-only (MPS falls back to no mixed precision); MPS OOM
 handling.
   - **`lora_trainer_tasks.py`** — failure guidance now mentions MPS (not just
 CUDA).
   - **`README.md`** — documents the Apple Silicon (MPS) training path.

   ## Notes

   - The Z-Image backend runs in `backend/venv` (MPS build of torch); the
 SDXL-legacy backend still requires `venv-torch/` (CUDA wheels), so on Apple
 Silicon the Z-Image backend is the practical path.
   - Refs #43

### CODE REVIEW

**Review by guaardvark — COMMENTED** [review]
Read through all five files. The CUDA path is unchanged in behaviour: `_dev()` resolves to `cuda` wherever the code used the literal, the probe still passes here (ran your new probe line under `backend/venv` on a CUDA box: `OK` / `cuda`), and the only new work on NVIDIA is that the every-25-steps hygiene now goes through `_cuda_reclaim()` (adds `gc.collect()` + `ipc_collect()`), which is fine at that cadence. `PYTORCH_ENABLE_MPS_FALLBACK` set on every platform is harmless. Nothing to change in the code.

What I need before merging, same bar as M1: evidence from one completed run on the Mac, since I cannot exercise MPS here.

1. The trainer's stderr from a real Z-Image LoRA job: the `staging VAE on mps` / `staging text encoder on mps` / `staging transformer+LoRA on mps` lines, the step/loss lines, and the save line, plus wall time and which machine (chip, RAM).
2. The clamp line if it printed. With `recommended_max_memory()` a 16 GB Mac lands around 10-11 GB, so it will train at 512; a 64 GB machine gets 768. Say which you got.
3. One image generated with the resulting LoRA loaded through the normal Images page, so the Kohya-style save is confirmed to load on the MPS pipeline.
4. macOS and torch versions, since `_is_bf16_supported()` answers True for any MPS device and bf16 on Metal depends on the OS/torch pairing.

Paste those here and I merge on the next green run.

**Comment by kwiksher** [comment]
**M2 evidence — completed MPS training run, plus where the timeout fixes live**

### 1. Timeout changes moved into this PR

The timeout adjustments that were bundled into the G3 RunPod snapshot now live here, since a Mac run cannot finish without them:

- `plugins/lora_trainer/real_trainer.py` — `_LOAD_TIMEOUT_S` 900 → 3600, `_TRAIN_TIMEOUT_S` 1800 → 10800
- `backend/tasks/lora_trainer_tasks.py` — `train_lora` `soft_time_limit=14700` / `time_limit=15300`; `reap_stuck_training` cutoff 45 → 270 min

They have been removed from `pr/g3-runpod-lora-trainer` (force-pushed `80da48a4` → `68fe8451`) so the two PRs don't overlap.

### 2. Fresh completed run on this PR's code (`4f902ba2`)

Apple M5 Pro, 48 GB · macOS 26.6.2 (25G83) · torch 2.13.0 · diffusers 0.39.0 · peft 0.20.0 · `torch.backends.mps.is_available() == True` · `recommended_max_memory()` = 37.44 GB

Subject "Elara", 11 training images, **400 steps**, rank 16, bf16, 768 res:

```
[run_zimage_trainer] loaded on CPU from .../Tongyi-MAI--Z-Image-Turbo dtype=torch.bfloat16
[run_zimage_trainer] staging VAE on mps for latent cache...
[run_zimage_trainer] cached 11 latents @ 768; VAE off GPU
[run_zimage_trainer] staging text encoder on mps for prompt cache...
[run_zimage_trainer] cached 11 prompts; TE off GPU
[run_zimage_trainer] staging transformer+LoRA on mps (res=768, rank=16, dtype=torch.bfloat16)...
[run_zimage_trainer] step 1/400 loss=6.06250
[run_zimage_trainer] step 51/400 loss=3.51562
[run_zimage_trainer] step 101/400 loss=2.18750
[run_zimage_trainer] step 151/400 loss=0.66797
[run_zimage_trainer] step 201/400 loss=0.99609
[run_zimage_trainer] step 251/400 loss=0.43164
[run_zimage_trainer] step 301/400 loss=0.62109
[run_zimage_trainer] step 351/400 loss=0.50781
[run_zimage_trainer] step 400/400 loss=0.68750
[run_zimage_trainer] saved LoRA → .../data/training/loras/Elara_v2.safetensors (33457088 bytes)
```

**Wall time: 4514 s (75.2 min)** — cold load + staging + 400 steps + save; ~11 s/step at 768.

**Clamp line:** none printed — 48 GB → `max_res` 768, and the job requested 768, so nothing to clamp.

### 3. Image with the trained LoRA — follows with M3/M4

M2 ships **training only**. On Apple Silicon the offline Z-Image inference path was CUDA-only, so rendering a still with the trained LoRA needs the next PRs in the track:

- **M3 (`pr/m3-zimage-mps-native`)** — native MPS support for the offline diffusers pipeline (no ComfyUI)

    > #183   created as a draft PR

- **M4 (`pr/m4-comfyui-image`)** — the alternative ComfyUI route

Both will be submitted **after M2 is merged**. `Elara_v2.safetensors` is ready to render with as soon as either lands.

**Comment by guaardvark** [comment]
Thanks for the complete run and the log; that is exactly the evidence I was after, and the CUDA path review above still stands. One change before this merges, then it goes in on green and M3 (#183) is next.

**The long timeouts need to be Mac-only.** As the PR stands they change the stock NVIDIA install too:

- `plugins/lora_trainer/real_trainer.py`: `_LOAD_TIMEOUT_S` 900 → 3600, `_TRAIN_TIMEOUT_S` 1800 → 10800
- `backend/tasks/lora_trainer_tasks.py`: `train_lora` gains `soft_time_limit=14700` / `time_limit=15300`, and `reap_stuck_training_subjects` moves from 45 to 270 minutes

On CUDA the 30 min train cap and the 45 min reaper were measured against real runs: a healthy Z-Image subject finishes well inside 30 min, so a run that is still going at 30 min is wedged, and the reaper frees the Subject 15 minutes later. With this PR as written, a wedged trainer on an NVIDIA box holds the GPU for 3 hours and a dead one leaves the Subject stuck in `training` for 4.5 hours. I would rather not change that for every stock user to make Apple Silicon work.

What I am asking for:

- One platform flag decided once, next to the probe you already added (`sys.platform == "darwin"` is enough; the MPS probe result is fine too if you prefer it), and every one of the four numbers derived from it. Stock values stay exactly as on `main` when the flag is off: 900 / 1800, the task keeps the global `task_time_limit`, reaper stays 45 minutes. Your Mac values apply when it is on.
- The reaper cutoff is a SQL interval string today; build it from the same constant so the three numbers cannot drift apart.
- Comments that say what the Mac numbers were measured against (your 75 min / 400 step run is a good anchor) and the README line saying these limits are Apple Silicon only.
- No new test needed for this; if you want one, assert the module constants for the non-Darwin case so a future edit cannot silently widen the CUDA limits again.

Everything else in the PR is fine as is.

**Comment by kwiksher** [comment]
Update pushed to pr/m2-lora-trainer-mps (f3192b05) — the long timeouts are now Apple-Silicon only.

 Thanks for the review. All four points are addressed.

 ### One platform flag, one derivation

 IS_APPLE_SILICON now lives next to the platform probes in backend/utils/platform.py (via the existing
 is_macos()), and all four numbers are derived from it through a pure helper:

 ```python
   IS_APPLE_SILICON = is_macos()

   def lora_train_time_budgets(apple_silicon):
       load = 3600 if apple_silicon else 900
       train = 10800 if apple_silicon else 1800
       if apple_silicon:
           task_soft = load + train + 300   # 245 min
           task_hard = load + train + 900   # 255 min
           reap = task_hard + 900           # 270 min
       else:
           task_soft = task_hard = None     # global task_time_limit applies
           reap = 45 * 60
       return load, train, task_soft, task_hard, reap
 ```

 ```┌──────────────────┬────────────────────────────────────────────────┬─────────────────┐
 │                  │ Stock CUDA (unchanged from main)               │ Apple Silicon   │
 ├──────────────────┼────────────────────────────────────────────────┼─────────────────┤
 │ daemon load cap  │ 900 s                                          │ 3600 s          │
 ├──────────────────┼────────────────────────────────────────────────┼─────────────────┤
 │ daemon train cap │ 1800 s                                         │ 10800 s         │
 ├──────────────────┼────────────────────────────────────────────────┼─────────────────┤
 │ Celery soft/hard │ none — global task_time_limit (2400 s) applies │ 14700 / 15300 s │
 ├──────────────────┼────────────────────────────────────────────────┼─────────────────┤
 │ reaper cutoff    │ 45 min                                         │ 270 min         │
 └──────────────────┴────────────────────────────────────────────────┴─────────────────┘
```
 - plugins/lora_trainer/real_trainer.py: the class caps now read LORA_LOAD_TIMEOUT_S / LORA_TRAIN_TIMEOUT_S.
 - backend/tasks/lora_trainer_tasks.py: soft_time_limit / time_limit are only passed when the flag is on, so
   stock keeps the global limits.

 ### Reaper built from the same constant

 The SQL no longer hardcodes an interval string — it binds LORA_REAP_STUCK_AFTER_S (now() -
 make_interval(secs => :stale_s)), so the reaper cutoff, the daemon caps, and the task limits cannot drift
 apart.

 ### Comments + README

 The Mac budget is anchored to the real 75 min / 400-step MPS run (and the Elara 151/640 failure), and the
 README states the longer timeouts are Apple Silicon only.

**Comment by guaardvark** [comment]
Merged. The CUDA budgets read exactly as on main (900 s load, 1800 s train, global Celery limits, 45 min reaper) and the Mac numbers now come from one place, which is better than what either of us had before. Thank you for carrying it through. #183 (M3) is next in line.
